cirq/protocols/approx_equality.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+from typing import Any
+
+from typing_extensions import Protocol
+
+
+class SupportsApproxEquality(Protocol):
+    """Object which can be approximately compared."""
+
+    def _approx_eq_(self, other: Any, rel_tol: float, abs_tol: float) -> bool:
+        """Approximate comparator.
+
+        Types implementing this protocol define their own logic for approximate
+        comparison with other types.
+
+        Args:
+        other: Target object for approximate comparison.
+        rel_tol: The relative tolerance. See math.isclose() documentation for
```

Comment:
Hmmmm, numpy uses `rtol` and `atol` but `math.isclose` uses `rel_tol` and `abs_tol`. I think for now we should stick to `rtol` and `atol`, and point at numpy's isclose. We can switch this later, to keep the whole library consistent.

Reply:
Before I saw your comment I already implemented backward-compatible `_isclose()`. It fallbacks to `math.isclose()` for Python >= 3.5 and` numpy.isclose()` wht Python is older.

I think this is strictly better, since `math.isclose()` usage is much cleaner (to make `numpy.isclose()` symmetric, I needed to implement some additional logic, see my code). I don't think we'll be able to switch to `math.isclose()` very soon, since it requires pretty new Python dependency. Also, `math.isclose()` is more strict then `numpy.isclose()`: if some test passes on Python >= 3.5 it will pass on Python < 3.5 (up to NaN and other special values handling which I didn't check for differences).

Reply:
I think we should be consistent between the two versions of python. If a method doesn't exist in python 2, and doesn't translate in a way that can be handled by the transpilation scripts, then we should avoid using that method. The only place I'm happy to condition on python 3 or not is in the tests.

Reply:
Sounds reasonable. I'll remove `math.isclose` completely in that case. Do you think it's reasonable to keep my wrapper around `numpy.isclose` to make it symmetric?

Reply:
Done.